crates/sources/src/sources/uniswap_v3/pool_fetching.rs

```diff
@@ -1,57 +1,231 @@
-use super::graph_api::{RegisteredPools, Token, UniV3SubgraphClient};
-use crate::recent_block_cache::Block;
+use super::graph_api::{PoolData, UniV3SubgraphClient};
 use crate::token_pair::TokenPair;
-use anyhow::Result;
-use ethcontract::{H160, U256};
-use num::BigInt;
+use anyhow::{Context, Result};
+use ethcontract::H160;
+use itertools::{Either, Itertools};
 use reqwest::Client;
-use std::collections::{HashMap, HashSet};
+use std::{
+    collections::{HashMap, HashSet},
+    sync::{Arc, Mutex, Weak},
+    time::{Duration, Instant},
+};
 
 #[async_trait::async_trait]
 pub trait PoolFetching: Send + Sync {
-    async fn fetch(&self, token_pairs: HashSet<TokenPair>, at_block: Block) -> Result<Vec<Pool>>;
+    async fn fetch(&self, token_pairs: &HashSet<TokenPair>) -> Result<Vec<PoolData>>;
 }
 
-pub struct Pool {
-    pub address: H160,
-    pub token0: Token,
-    pub token1: Token,
-    pub sqrt_price: U256,
-    pub liquidity: U256,
-    pub tick: BigInt,
-    pub liquidity_net: HashMap<BigInt, BigInt>,
-    pub fee_tier: U256,
+pub struct CachedPool {
+    pub pool: PoolData,
+    pub updated_at: Instant,
+    pub requested_at: Instant,
 }
 
 pub struct UniswapV3PoolFetcher {
-    pub graph_api: UniV3SubgraphClient,
-    pub registered_pools: RegisteredPools,
+    graph_api: UniV3SubgraphClient,
+    pools_by_token_pair: HashMap<TokenPair, HashSet<H160>>,
+    cache: Mutex<HashMap<H160, CachedPool>>,
+    max_age: Duration,
 }
 
 impl UniswapV3PoolFetcher {
-    pub async fn new(chain_id: u64, client: Client) -> Result<Self> {
+    /// Retrieves all registered pools on Uniswap V3 subgraph, but without `ticks`,
+    /// making the cache values outdated immediately. Cache values are supposed to be updated
+    /// either on fetch or on periodic maintenance update.
+    pub async fn new(chain_id: u64, max_age: Duration, client: Client) -> Result<Self> {
         let graph_api = UniV3SubgraphClient::for_chain(chain_id, client)?;
+        let registered_pools = graph_api.get_registered_pools().await?;
+        tracing::debug!(
+            block = %registered_pools.fetched_block_number, pools = %registered_pools.pools.len(),
+            "initialized registered pools",
+        );
+
+        let mut pools_by_token_pair: HashMap<TokenPair, HashSet<H160>> = HashMap::new();
+        for pool in registered_pools.pools {
+            let token0 = pool.token0.clone().context("token0 does not exist")?.id;
+            let token1 = pool.token1.clone().context("token1 does not exist")?.id;
```

Comment:
This is existing code but how come the tokens are optional in this struct?

Reply:
It's because the `PoolData` struct is used for two different calls on subgraph where in one case tokens are fetched while on other not.